### PR TITLE
Terms aggregation should remap global ordinal buckets when a sub-aggregator is used to sort the terms

### PR DESCRIPTION
**Pull request by jimczi**
`terms` aggregations at the root level use the `global_ordinals` execution hint by default.
When all sub-aggregators can be run in `breadth_first` mode the collected buckets for these sub-aggs are dense (remapped after the initial pruning).
But if a sub-aggregator is not deferrable and needs to collect all buckets before pruning we don't remap global ords and the aggregator needs to deal with sparse buckets.
Most (if not all) aggregators expect dense buckets and uses this information to allocate memories.
This change forces the remap of the global ordinals but only when there is at least one sub-aggregator that cannot be deferred.

Relates #24788

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java [hunk 1]
```diff
@@ -254,109 +294,54 @@ protected final XContentBuilder keyToXContent(XContentBuilder builder) throws IO
...
         @Override
-        protected LeafBucketCollector newCollector(final SortedSetDocValues ords,
-                LeafBucketCollector sub) {
-            segmentDocCounts = context.bigArrays().grow(segmentDocCounts, 1 + ords.getValueCount());
+        public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                    final LeafBucketCollector sub) throws IOException {
+            if (segmentOrds != null) {
+                mapSegmentCountsToGlobalCounts();
+            }
+            globalOrds = valuesSource.globalOrdinalsValues(ctx);
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java]: why is this needed, I don't see where its used?

**Review by colings86 — APPROVED** [review]
LGTM, I left one comment but only for my own understanding, I don't think it will affect whether you merge this as is.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java (hunk 1 above)
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java]: It is used in https://github.com/jimczi/elasticsearch/blob/5b79cbe87c5c28cafed068a3910e1298a9dd700d/core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java#L385 to remap the segment ordinals collected in this segment to global ordinals.

**Comment by IdanWo** [comment]
Up?
